RFCs/0033-graph-query.md

```diff
@@ -0,0 +1,491 @@
+
+In the context of this RFC, the notion of a PartiQL value has been extended,
+according to RFC-0025, to include graph values, while the notion of an expression
+is being extended by adding graph matching expressions.
+Furthermore, a graph matching expression can contain PartiQL expressions,
+notably within filtering `WHERE` clauses.
+They, in turn, can contain graph-specific predicates like `IS DIRECTED` and
+`ALL_DIFFERENT`.
+
+A PartiQL expression is evaluated in an environment that maps variables to PartiQL values,
+with the result of an evaluation being a PartiQL value.  
+The above-outlined evaluation semantics of graph pattern matching in GPML is
+structured differently (for one, it produces a bag of tuples of annotated paths)
+and operates over values of a kind that are not part
+of PartiQL data model, even after the extension (such as nodes and edges
+that have awareness of their identity and location within a graph).
+In a sense, graph matching is a distinct query language of its own, embedded
+within PartiQL as a `MATCH` expression form.
+
+Consequently, our goal here is to describe how the two semantics are combined,
+specifically addressing:
+- Role of PartiQL environments while evaluating pattern matching.
+- Evaluation of PartiQL expressions within graph `WHERE` clauses.
+- And, crucially, how the result of pattern match computation is
+  represented as a value within the PartiQL data model.
+
+
+For evaluating a  graph matching expression
+>    `(` _graph_ `MATCH` _pattern_ `)`
+
+this RFC  proposes the following.
+
+- As any PartiQL expression, the graph matching expression
+  is evaluated within a regular PartiQL environment (𝝆<sub>0</sub>, 𝝆) that maps variables
+  (defined outside this expression) to values.
+
+- The _graph_ subexpression is evaluated within (𝝆<sub>0</sub>, 𝝆) and must result
+  in a graph value.
+
+- Given the graph value, graph matching computation for _pattern_ proceeds according
+  to the GPML semantics, with the following modifications:
+
+  - Variables defined in _pattern_ shadow variables bound in (𝝆<sub>0</sub>, 𝝆).
+
+  - Whenever a `WHERE e` clause within the pattern needs to be evaluated,
+    the clause's expression `e` is evaluated within (𝝆<sub>0</sub>, 𝝆)
+    though regular PartiQL evaluation, except that any pattern-defined
+    variable `x` is not looked up in (𝝆<sub>0</sub>, 𝝆).
+    Instead, its value, as a graph element, is provided by the GPML
+    pattern matching procedure.  
+    The subsequent processing of the graph-element value depends on the
+    context of `x`:
+    - If `x` is used within a graph-specific construct (such as a predicate
+      like `IS DIRECTED` or a path-aggregating function), the graph-element
+      value is passed to this construct directly.
+    - Otherwise, `x` is used as a regular PartiQL value.
+      - If `x` is bound to a node or an edge, the payload value of the latter
+        is used.
+      - If `x` is bound to a path, it is an error.
```

Comment:
Given our recent discussion offline about this, I think we should put this as an open question.  Specifically, these graph operations on variables that require _referencing_ the graph node/edge in its context (and not just the value associated with the node/edge) is a pretty significant question to answer.

If we do not need to support it in the initial RFC, we can defer this.  Specifically, this defers the notion of modeling a value that is a _graph reference_ (i.e., the tuple `<id, G>` where `id` is the node/edge identifier in some graph `G`).  I think we _could_ introduce this concept as a first-class type, but then we have a lot of rules about when the value is used (i.e., de-referencing) versus the reference itself.  An option to put in the open question is if this type of value is second-class and only available to certain predicates in a special GPML environment.  Such a definition could make it less general, but probably sufficient for most use cases.

Reply:
Done: 3ace887